src/ui/DynamicFacetSearch/DynamicFacetSearch.ts

```diff
@@ -0,0 +1,58 @@
+import 'styling/DynamicFacetSearch/_DynamicFacetSearch';
+import { $$ } from '../../utils/Dom';
+import { FacetSearchController } from '../../controllers/FacetSearchController';
+import { DynamicFacet } from '../DynamicFacet/DynamicFacet';
+import { IFacetSearchResponse } from '../../rest/Facet/FacetSearchResponse';
+import { Utils } from '../../utils/Utils';
+import { DynamicFacetSearchInput } from './DynamicFacetSearchInput';
+
+export class DynamicFacetSearch {
+  public element: HTMLElement;
+  private input: DynamicFacetSearchInput;
+  private facetSearchController: FacetSearchController;
+  private facetSearchTimeout: number;
+  private facetSearchPromise: Promise<IFacetSearchResponse>;
+
+  constructor(private facet: DynamicFacet) {
+    this.element = $$('div', { className: 'coveo-dynamic-facet-search' }).el;
+    this.createAndAppendInput();
+    this.facetSearchController = new FacetSearchController(this.facet);
+  }
+
+  private createAndAppendInput() {
+    this.input = new DynamicFacetSearchInput(this.facet, this.onInputChange.bind(this));
+    this.element.appendChild(this.input.element);
+  }
+
+  private onInputChange(value: string) {
+    this.cancelAnyPendingSearchOperation();
+
+    if (Utils.isEmptyString(value)) {
+      return;
+    }
+
+    this.facetSearchTimeout = window.setTimeout(() => {
```

Comment:
Instead of delay, I would try to change that to a debounced function with a trailing call (versus head call).

Reply:
Not sure I understand what you mean about trailing call vs head call? the search is only performed after the delay.
Debounced is pretty much the same implementation has what is done here, but maybe it would look a bit cleaner. 😄